javascript/ql/lib/semmle/javascript/NPM.qll

```diff
@@ -12,8 +12,25 @@ class PackageJson extends JsonObject {
     this.isTopLevel()
   }
 
-  /** Gets the name of this package. */
-  string getPackageName() { result = this.getPropStringValue("name") }
+  /**
+   * Gets the name of this package.
+   * If the package is located under the package `pkg1` and its relative path is `foo/bar`, it can be `pkg1/foo/bar`.
+   */
+  string getPackageName() {
+    result = this.getPropStringValue("name")
+    or
+    exists(
+      PackageJson parentPackage, string currentDir, string parentDir, string parentPackageName
+    |
+      currentDir = this.getJsonFile().getParentContainer().getAbsolutePath() and
+      parentDir = parentPackage.getJsonFile().getParentContainer().getAbsolutePath() and
+      parentPackageName = parentPackage.getPropStringValue("name") and
+      parentDir.indexOf("node_modules") != -1 and
+      currentDir != parentDir and
+      currentDir.indexOf(parentDir) = 0 and
```

Comment:
You are doing a string-comparison here, and I'm not a fan of that.  

You could instead drop the `.getAbsolutePath()`, and then do something like: 
`parentDir.getAChildContainer+() = currentDir`

Reply:
Thank you!
That looks better!
I refactored the codes here.